app/process_proposal.go

```diff
@@ -0,0 +1,79 @@
+package app
+
+import (
+	"github.com/celestiaorg/celestia-app/x/payment/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+const (
+	rejectedPropBlockLog = "Rejected proposal block:"
+)
+
+func (app *App) ProcessProposal(req abci.RequestProcessProposal) abci.ResponseProcessProposal {
+	// Check for message inclusion:
+	//  - each MsgPayForMessage included in a block should have a corresponding message also in the block data
+	//  - the commitment in each PFM should match that of its corresponding message
+	//  - there should be no unpaid for messages
+
+	// extract the commitments from any MsgPayForMessages in the block
+	commitments := make(map[string]struct{})
+	for _, rawTx := range req.BlockData.Txs {
+		tx, err := app.txConfig.TxDecoder()(rawTx)
+		if err != nil {
+			continue
+		}
+
+		for _, msg := range tx.GetMsgs() {
+			if sdk.MsgTypeURL(msg) != types.URLMsgPayforMessage {
+				continue
+			}
+
+			pfm, ok := msg.(*types.MsgPayForMessage)
+			if !ok {
+				continue
+			}
+
+			commitments[string(pfm.MessageShareCommitment)] = struct{}{}
+		}
+
+	}
```

Comment:
I don't see any reason to forbid including two of the same message in the same block, so long as they're paid for. _That being said_, a block proposer could be tricky and make both PayForMessage txs in the block point to the same message! An interesting corner case. Can you file an issue...somewhere...to re-visit this? And document your decision (forbid or allow, doesn't really matter, impl detail) in the code with a link to the issue.

Reply:
#226 

will update this PR tmrw to make the default behavior to exclude the identical message, but keep the payments

Reply:
update to this, I just ended up taking the easier option, in which we count the commitments using a counter instead of relying on a map, this way we also count identical commitments and I think things work intuitively. Intuitively being that identical messages get included twice and each PayForData must have a corresponding message. 

we should still make a formal decision with #226